dm/master/shardddl/pessimist_test.go

```diff
@@ -279,3 +287,518 @@ func (t *testPessimist) TestPessimist(c *C) {
+	var (
+		watchTimeout  = 500 * time.Millisecond
+		task          = "task-source-re-entrant"
+		source1       = "mysql-replica-1"
+		source2       = "mysql-replica-2"
+		source3       = "mysql-replica-3"
+		schema, table = "foo", "bar"
+		DDLs          = []string{"ALTER TABLE bar ADD COLUMN c1 INT"}
+		ID            = fmt.Sprintf("%s-`%s`.`%s`", task, schema, table)
+		i11           = pessimism.NewInfo(task, source1, schema, table, DDLs)
+		i12           = pessimism.NewInfo(task, source2, schema, table, DDLs)
+		i13           = pessimism.NewInfo(task, source3, schema, table, DDLs)
+
+		sources = func(task string) []string {
+			switch task {
+			case task:
+				return []string{source1, source2, source3}
+			default:
+				c.Fatalf("unsupported task %s", task)
+			}
+			return []string{}
+		}
+		logger = log.L()
+		p      = NewPessimist(&logger, sources)
+	)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// 0. start the pessimist.
+	c.Assert(p.Start(ctx, etcdTestCli), IsNil)
+	c.Assert(p.Locks(), HasLen, 0)
+	defer p.Close()
+
+	// 1. PUT i11 and i12, will create a lock but not synced.
+	_, err := pessimism.PutInfo(etcdTestCli, i11)
+	c.Assert(err, IsNil)
+	_, err = pessimism.PutInfo(etcdTestCli, i12)
+	c.Assert(err, IsNil)
+	c.Assert(utils.WaitSomething(30, 100*time.Millisecond, func() bool {
+		lock := p.Locks()[ID]
+		if lock == nil {
+			return false
+		}
+		_, remain := lock.IsSynced()
+		return remain == 1
+	}), IsTrue)
+
+	// 2. re-PUT i11, to simulate the re-entrant of the owner before the lock become synced.
+	rev1, err := pessimism.PutInfo(etcdTestCli, i11)
+	c.Assert(err, IsNil)
+
+	// 3. re-PUT i12, to simulate the re-entrant of the non-owner before the lock become synced.
+	rev2, err := pessimism.PutInfo(etcdTestCli, i12)
+	c.Assert(err, IsNil)
+
+	// 4. wait exec operation for the owner become available.
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		opCh := make(chan pessimism.Operation, 10)
+		ctx2, cancel2 := context.WithTimeout(ctx, watchTimeout)
+		pessimism.WatchOperationPut(ctx2, etcdTestCli, task, source1, rev1, opCh)
+		cancel2()
+		close(opCh)
+		c.Assert(len(opCh), Equals, 1)
+		op := <-opCh
+		c.Assert(op.Exec, IsTrue)
+		c.Assert(op.Done, IsFalse)
+	}()
+
+	// 5. put i13, the lock will become synced, then an operation PUT for the owner will be triggered.
+	_, err = pessimism.PutInfo(etcdTestCli, i13)
+	c.Assert(err, IsNil)
+	wg.Wait()
+
+	// 6. re-PUT i11, to simulate the re-entrant of the owner after the lock become synced.
+	rev1, err = pessimism.PutInfo(etcdTestCli, i11)
+	c.Assert(err, IsNil)
+
+	// 8. wait exec operation for the owner become available again (with new revision).
+	opCh := make(chan pessimism.Operation, 10)
+	ctx2, cancel2 := context.WithTimeout(ctx, watchTimeout)
+	pessimism.WatchOperationPut(ctx2, etcdTestCli, task, source1, rev1, opCh)
+	cancel2()
+	close(opCh)
+	c.Assert(len(opCh), Equals, 1)
+	op11 := <-opCh
+	c.Assert(op11.Exec, IsTrue)
```

Comment:
I don't quite understand here. Why do we get `exec=true` again if we re-PUT i11?